src/vsr/journal.zig

```diff
@@ -703,83 +843,505 @@ pub fn Journal(comptime Replica: type, comptime Storage: type) type {
+            assert(self.recovering);
+            // We expect that no other process is issuing reads while we are recovering.
+            assert(self.reads.executing() == 0);
+
+            if (slot.index == slot_count) {
+                self.recover_done();
+                return;
+            }
+            assert(slot.index < slot_count);
+
+            const message = replica.message_bus.get_message();
+            defer replica.message_bus.unref(message);
+
+            const read = self.reads.acquire() orelse unreachable;
+            read.* = .{
+                .self = self,
+                .completion = undefined,
+                .message = message.ref(),
+                .callback = undefined,
+                .op = undefined,
+                .checksum = slot.index,
+                .destination_replica = null,
+            };
+
+            log.debug("{}: recover_prepares: recovering slot={}", .{
+                self.replica,
+                slot.index,
+            });
+
+            self.storage.read_sectors(
+                recover_prepares_on_read,
+                &read.completion,
+                // Only the header of the message needs to be read (128 bytes).
+                // However, the minimum read is a 4 KiB sector.
+                message.buffer[0..config.sector_size],
+                offset_physical(.prepares, slot),
+            );
+        }
+
+        /// When in doubt about whether a particular message was received, it must be marked as
+        /// faulty to avoid nacking a prepare which was received then lost/misdirected/corrupted.
+        ///
+        ///
+        /// During recovery, the in-memory header of a faulty slot must be left reserved.
+        /// The recovery process must be conservative about which headers are stored in
+        /// `journal.headers`. To understand why this is important, consider what happens if it
+        /// did load the faulty header into `journal.headers`:
+        ///
+        /// 1. Suppose slot 8 is in case @E or @F. Per the table below, mark slot 8 faulty.
+        /// 2. Suppose slot 9 is also loaded as faulty.
+        /// 3. Journal recovery finishes. The replica beings to repair its missing/broken messages.
+        /// 4. VSR recovery protocol fetches the true prepare for slot 9.
+        /// 5. The message from step 4 is written to slot 9 of the prepares.
+        /// 6. The header from step 4 is written to slot 9 of the redundant headers.
+        ///    But writes to the redundant headers are done in batches of `headers_per_sector`!
+        ///    So if step 1 loaded slot 8's prepare header into `journal.headers`, slot 8's
+        ///    redundant header would be updated at the same time (in the same right) as slot 9.
+        /// 7! Immediately after step 6's write finishes, suppose the replica crashes (e.g. due to
+        ///    power failure.
+        /// 8! Journal recovery again — but now slot 8 is loaded *without* being marked faulty.
```

Comment:
I remember talking about this with you, but how would step 8 here then lead into a correctness bug? Would we not detect the header/prepare mismatch and still mark faulty?

What would be the next step here to cause data loss? Would we not detect any mismatch between header/prepare as faulty?

Reply:
I know that there is something here, I just can't remember it 😄

Reply:
> Would we not detect the header/prepare mismatch and still mark faulty?

You're right. I think our conversation was when `@E` and `@F` were set to `fix` instead of `vsr`.

But I think its still an issue for case `@D` (rather than `@E`/`@F`), since case `@D` could degrade to `@H` after the "repair"/restart (which has a different decision). In that case, we may wind up incorrectly nacking the op, since `@H` doesn't mark faulty.